velox/exec/Window.cpp

```diff
@@ -282,6 +314,131 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
   }
 }
 
+void Window::updateKRowsFrameBounds(
+    const bool isKPreceding,
+    const FrameChannelArg& frameArg,
+    const vector_size_t startRow,
+    const vector_size_t numRows,
+    vector_size_t* rawFrameBounds) {
+  auto firstPartitionRow = partitionStartRows_[currentPartition_];
+
+  if (frameArg.index == kConstantChannel) {
+    auto constantOffset = frameArg.constant.value();
+    std::iota(
+        rawFrameBounds,
+        rawFrameBounds + numRows,
+        startRow + (isKPreceding ? -constantOffset : constantOffset) -
+            firstPartitionRow);
+  } else {
+    frameArg.value->resize(numRows);
+    windowPartition_->extractColumn(
+        frameArg.index, partitionOffset_, numRows, 0, frameArg.value);
+    auto offsets = frameArg.value->values()->as<int64_t>();
+    for (auto i = 0; i < numRows; i++) {
+      if (offsets[i] <= 0) {
+        VELOX_USER_FAIL(
```

Comment:
It would be nice to make this check consistent with similar check for constant value. Also, we need to check for null values.

```
    for (auto i = 0; i < numRows; i++) {
      VELOX_USER_CHECK(!frameArg.value->isNullAt(i), "k in frame bounds cannot be null");
      VELOX_USER_CHECK_GE(
          offsets[i], 1, "k in frame bounds must be at least 1");
    }
```